src/shared/util.js

```diff
@@ -1218,20 +1218,108 @@ function MessageHandler(sourceName, targetName, comObj) {
   this.sourceName = sourceName;
   this.targetName = targetName;
   this.comObj = comObj;
-  this.callbackIndex = 1;
+  this.callbackId = 1;
+  this.streamId = 1;
   this.postMessageTransfers = true;
-  var callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
-  var ah = this.actionHandler = Object.create(null);
+  let streamSinks = this.streamSinks = Object.create(null);
+  let streamControllers = this.streamControllers = Object.create(null);
+  let callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
+  let ah = this.actionHandler = Object.create(null);
 
   this._onComObjOnMessage = (event) => {
-    var data = event.data;
+    let data = event.data;
     if (data.targetName !== this.sourceName) {
       return;
     }
-    if (data.isReply) {
-      var callbackId = data.callbackId;
+    if (data.stream) {
+      let sourceName = this.sourceName;
+      let targetName = data.sourceName;
+      let streamId = data.streamId;
+      let sendStreamResponse = ({ stream, success }) => {
```

Comment:
here and at sendStreamRequest add 'reason' parameter after success -- it shall be populated/used when success == false. reject on xxx_complete side shall accept this reason as argument.